### ISSUE TITLE
Client-facing connections get closed despite using errorfile 

### ISSUE THREAD

**Issue by jaroslawr** [haproxy/haproxy#891]
I am not able to get haproxy to keep the client-facing connection alive when the backend is not accepting a connection or when it time outs. I use an errorfile with `http/1.1` and no 'Connection: close' but haproxy seems to add its own 'Connection: close' header.

I use haproxy 2.2.3:

```
HA-Proxy version 2.2.3-6a551a-16 2020/09/23 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.2.3.html
Running on: Linux 5.6.0-1-amd64 #1 SMP Debian 5.6.7-1 (2020-04-29) x86_64
```

My haproxy.cfg:

```
defaults
        mode http
        timeout client 10000
        timeout connect 100
        timeout server 5000

frontend lolcathost
        default_backend jetty
        bind localhost:8080

backend jetty
        errorfile 503 204.http
        errorfile 504 204.http
        server localhost_8082 localhost:8082
```

204.http is this:

```
HTTP/1.1 204 No Content
Cache-Control: no-cache

```

If the server defined in jetty backend is up the connection is reused correctly:

```
curl -v -v localhost:8080 localhost:8080
*   Trying ::1:8080...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< date: Wed, 07 Oct 2020 19:18:27 GMT
< content-type: text/html;charset=utf-8
< content-length: 21
<
<h1>Hello World</h1>
* Connection #0 to host localhost left intact
* Found bundle for host localhost: 0x55c920fd7330 [serially]
* Can not multiplex, even if we wanted to!
* Re-using existing connection! (#0) with host localhost
* Connected to localhost (::1) port 8080 (#0)
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< date: Wed, 07 Oct 2020 19:18:27 GMT
< content-type: text/html;charset=utf-8
< content-length: 21
<
<h1>Hello World</h1>
* Connection #0 to host localhost left intact
```

If you bring up haproxy but the backend server is not up, haproxy will use 204.http but "connection: close" is added:

```
curl -v -v localhost:8080 localhost:8080
*   Trying ::1:8080...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 204 No Content
< cache-control: no-cache
< connection: close
<
* Closing connection 0
* Hostname localhost was found in DNS cache
*   Trying ::1:8080...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#1)
> GET / HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 204 No Content
< cache-control: no-cache
< connection: close
<
* Closing connection 1
```

Same thing happens when the jetty backend takes above `timeout server` to respond.

**Comment by capflam** [comment]
So I confirm the bug for 204/304 responses. While these responses are correctly handled when received from the server, It is not true for errorfiles. Because there is neither content-length nor transfer-encoding, a "Connection: close" is automatically added when the response is sent to the client. Of course, 204/304 responses have no body and should not provide these headers. This case must be handled during errorfiles parsing.

I will produce a patch soon. In the mean time, a workaround is to add a "Content-Length: 0' header.

**Comment by capflam** [comment]
It should be fixed now in 2.3-DEV. I'll backport the fix soon. Thanks !

**Comment by capflam** [comment]
done !
